File: packages/execution-engine/src/models/signal.py
from enum import Enum

from typing import Optional

from pydantic import BaseModel, Field

from .timeframe import Timeframe


class SignalDirection(str, Enum):
    BUY = "BUY"
    SELL = "SELL"


class SignalMode(str, Enum):
    BACKTEST = "backtest"
    FORWARD_TEST = "forward_test"
    LIVE = "live"


class BOSType(str, Enum):
    BULLISH = "bullish"
    BEARISH = "bearish"


class SignalMetadata(BaseModel):
    bos_type: BOSType
    liquidity_swept: bool
    session: str
    spread_at_generation: float
    volatility_ratio: float


class Signal(BaseModel):
    id: str  # UUID
    instrument: str  # e.g. "US30"
    direction: SignalDirection
    entry_price: float
    stop_loss: float
    take_profit: float
    position_size: float  # lot size
    confidence_score: float = Field(ge=0.0, le=1.0)
    timeframe: Timeframe
    order_block_id: str
    strategy_id: str
    mode: SignalMode
    metadata: SignalMetadata
    exit_rules: Optional[dict] = None  # ExitRules config from strategy
    broker_symbol: Optional[str] = None  # resolved by backend gateway
    created_at: str  # ISO 8601

File: packages/backend/src/common/types/signal.ts
import { Timeframe } from './timeframe';

export type SignalDirection = 'BUY' | 'SELL';
export type SignalMode = 'backtest' | 'forward_test' | 'live';
export type BOSType = 'bullish' | 'bearish';

export interface SignalMetadata {
  bos_type: BOSType;
  liquidity_swept: boolean;
  session: string;
  spread_at_generation: number;
  volatility_ratio: number;
}

export interface Signal {
  id: string;                    // UUID
  instrument: string;            // e.g. "US30"
  direction: SignalDirection;
  entry_price: number;
  stop_loss: number;
  take_profit: number;
  position_size: number;         // lot size
  confidence_score: number;      // 0.0 - 1.0
  timeframe: Timeframe;
  order_block_id: string;
  strategy_id: string;
  mode: SignalMode;
  metadata: SignalMetadata;
  created_at: string;            // ISO 8601
  broker_symbol?: string;        // Resolved from AccountInstrument mapping
}
